### ISSUE TITLE
Save all done Tasks to a collection

### ISSUE THREAD

**Issue by tijlleenders** [tijlleenders/ZinZen#1407]
**Is your feature request related to a problem? Please describe.**
It would be very nice if the scheduler would take into account what I've already completed today, and in the past.

**Describe the solution you'd like**
- [ ] #2063
- the goalId
- the datetime of completion, with the scheduled startDatetime and duration
- (the datetime(s) of postpone)
- (the datetime(s) of skip)

- [ ] Add this info to the `tasksCompletedToday` object that is sent to the scheduler, so completed Tasks remain fixed on their slot for today (with a completed status), even when then scheduler inputs change.

**Describe alternatives you've considered**
Keep this state purely in the frontend, and don't send this to the scheduler.  
This doesn't work as the scheduler produces a different schedule whenever goals change, making the completed Tasks possibly appear at different times today, or even in the future instead of today.

**Additional context**
This collection can later also be used for habit or goal tracking reports.  
Required for ( https://github.com/tijlleenders/ZinZen-scheduler/issues/466 ) to be implemented.

**Discussion required**  
It is unclear to me what the `taskCollection` is used for. Is this just to store the specific Tasks that have been completed today, so we can send them to the Scheduler as `tasksCompletedToday`? If so then why index them with a Goal id?

Maybe there should be three collections related to Tasks?:
- schedulerOutputCache
Only refreshed if Goals change - or if a new day has started.
- tasksDoneToday
  These should be cleared when (to)day is past. Currently the completed Tasks from yesterday are still 'stuck' in this collection - and confuse the Scheduler when they are sent - as they are not __today__
- tasksHistory  
This is the collection referred to in this issue, which can later be used for statistics, indexed by goalId, as that would be the logical query pattern.

**Comment by tijlleenders** [comment]
@Tushar-4781 Please advise how to approach this issue, as you said in voice that there was already some provision for doing this?

**Comment by vinaybadgujar102** [comment]
> [@Tushar-4781](https://github.com/Tushar-4781) Please advise how to approach this issue, as you said in voice that there was already some provision for doing this?

@tijlleenders @Tushar-4781 Is there any follow-up on this? Please mention any specific details if there are any about the implementation of this feature.

**Comment by Tushar-4781** [comment]
i need to gather my thoughts on rust side like how is it going to accept the done hrs of every task. will have a disucssion on this. though the feature of saving the progress is already implemented. it would just be minor refactoring.

**Comment by tijlleenders** [comment]
@Tushar-4781 Ok all scheduler/Task related things are waiting on this to be clear.

**Comment by tijlleenders** [comment]
@Tushar-4781 How about now?

**Comment by tijlleenders** [comment]
Blocked by #1918

**Comment by tijlleenders** [comment]
Don't know why this was blocked. There seems to be no dependency between logging Tasks done and the ability to move Goals around in the hierarchy...

**Comment by tijlleenders** [comment]
@Tushar-4781 @vinaybadgujar102 
It is unclear to me what the `taskCollection` is used for. Is this just to store the specific Tasks that have been completed today, so we can send them to the Scheduler as `tasksCompletedToday`? If so then why index them with a Goal id?

Maybe there should be three collections related to Tasks?:
- schedulerOutputCache
Only refreshed if Goals change - or if a new day has started.
- tasksDoneToday
  These should be cleared when (to)day is past. Currently the completed Tasks from yesterday are still 'stuck' in this collection - and confuse the Scheduler when they are sent - as they are not __today__
- tasksHistory  
This is the collection referred to in this issue, which can later be used for statistics, indexed by goalId, as that would be the logical query pattern.

**Comment by vinaybadgujar102** [comment]
@tijlleenders,
The taskCollection is used to store tasksCompletedToday, as you mentioned. It also stores blockedSlots and skippedTasks.

We already have the "Dumpbox" collection to store the scheduler cache. To make things clearer, we can put the "done today" tasks in a separate collection, as you suggested.

**Comment by tijlleenders** [comment]
Sounds good.  

Also, let's rename 'Dumpbox' to 'schedulerCache'.

**Comment by vinaybadgujar102** [comment]
Hey @tijlleenders,
Just to confirm, the changes you requested are as follows:

1. Rename "Dumpbox" to "schedulerCache".
2. Create a "tasksDoneToday" collection, which will forward data to the Rust scheduler and will be cleared when day is over.
3. Create a "tasksHistory" collection to store the following events: Done, Postpone, Skip.

Right?

**Comment by tijlleenders** [comment]
@vinaybadgujar102 Yes, except 1. Should be "scheduler**Output**Cache"

**Comment by vinaybadgujar102** [comment]
@tijlleenders Could you provide example of how to store data for each each task event type ("completed," "postponed," and "skipped") in the taskHistory collection? Just looking for the structure.

**Comment by tijlleenders** [comment]
> @tijlleenders Could you provide example of how to store data for each each task event type ("completed," "postponed," and "skipped") in the taskHistory collection? Just looking for the structure.

The idea is to be able to be flexible in the type of reports we can generate in the future, so granularity is on the specific event level.  
`goalId; datetime; completed/postponed/skipped`  

Queryable/indexed by goalId as I don't think statistics as %completed vs skipped make any sense without the specific goalId.

**Comment by vinaybadgujar102** [comment]
```javascript
export type TaskHistoryEvents = "completed" | "postponed" | "skipped";

export interface TaskHistoryItem {
  id: string;
  goalId: string;
  eventType: TaskHistoryEvents;
  scheduledStart?: string;
  scheduledEnd?: string;
  duration?: number;
  eventTimestamp: string;
}
```

@tijlleenders, confirm if this taskHistory item structure is correct

**Comment by tijlleenders** [comment]
@vinaybadgujar102 I don't see any use for an id.  
For a separate issue on the backlog: `goalId` should be of uuid type - not string.  
`duration`/`scheduledEnd`are redundant - but OK for convenience.  
Why is `scheduledStart/scheduledEnd/duration` optional? I don't think we can have something completed/postponed/skipped if it doesn't have these fields.  

Queries will typically be 'give me stats for `goalId` X. If needs change we can iterate.

**Comment by vinaybadgujar102** [comment]
>  I don't see any use for an id, especially not a string id. Indexeddb will probably make an integer? id underwater anyway?

agreed. db can handleid internally

> Why is scheduledStart/scheduledEnd/duration optional? I don't think we can have something completed/postponed/skipped if it doesn't have these fields.

 i thought, in the case of a "postpone" event, we might need not need to store the scheduled time details as they are inrelevant for this event case.

**Comment by tijlleenders** [comment]
There might be a use for it... like when (what day of the week / time) am I most likely to postpone things?
